timemachine/fe/plots.py

```diff
@@ -322,3 +322,83 @@ def plot_fwd_reverse_predictions(
     plot_png = buffer.read()
 
     return plot_png
+
+
+def plot_hrex_swap_acceptance_rates(swap_acceptance_rates: NDArray):
```

Comment:
> One plot that may be "missing" is an all-pairs transition matrix.

By "all-pairs transition matrix" I'm assuming you mean a $K \times K$ matrix where the $(i,\,j)$ element is transition rate between state $i$ and state $j$? If so, I didn't think to include this because I'd been focused only on nearest-neighbor swaps in a 1-d chain, but agreed that this seems useful for more general situations. Will add.

Reply:
Yes, exactly. 

>I didn't think to include this because I'd been focused only on nearest-neighbor swaps in a 1-d chain, but agreed that this seems useful for more general situations. Will add.

If I'm understanding correctly, that would still be informative even if the "inner loop" swaps are nearest neighbor only. (Since, in the "outer loop" (K**4 inner loop steps), `i` can swap with `j` that's not immediately adjacent.)

Reply:
> If I'm understanding correctly, that would still be informative even if the "inner loop" swaps are nearest neighbor only. (Since, in the "outer loop" (K**4 inner loop steps), i can swap with j that's not immediately adjacent.)

Ah, interesting. As it is we don't separately track the transition rates for the "outer loop" moves, and I don't think this is recoverable from the individual swap move accepted/proposed counts. I think this could be inferred from the replica-to-state mappings that we currently do track, though. Might need to think about this a little more.

Reply:
Added the transition matrix for "outer loop", i.e. permutation moves to `HrexDiagnostics` and added a plot in https://github.com/proteneer/timemachine/pull/1128/commits/341c286d853c0d014955d8cc93d77525eac60feb.

Also added an estimate of the relaxation time computed from the second-largest eigenvalue to the diagnostics in https://github.com/proteneer/timemachine/pull/1128/commits/e43369992207c5cccbef1c97cc5cf65f9abd7b8b.

Reply:
Sweet! These will be highly informative about how effectively we're sampling `p(state_permutations | xs)`.

Tangent 1: I don't expect this to be a bottleneck, but in case it is, I've started playing around a little with a possibly faster-mixing alternative to the uniform mixture of nearest-neighbor swap moves. (Apply [locally-balanced discrete MCMC](https://gist.github.com/maxentile/5497a1a87cf51fb132962ca26462dc0a) using some definition of a permutation's "neighborhood", e.g. considering in one proposal step [the set of all permutations that could have been accessed in T successive neighbor swap moves](https://gist.github.com/maxentile/9a90704b6df4f8cd2e255a033ee25e10).)

Tangent 2: In the future, we might consider going all the way to the "infinite swap limit." (An approach introduced by [Dupuis+ (2012)](https://epubs.siam.org/doi/abs/10.1137/110853145), reformulated and adapted to MD context by [Lu+ (2013)](https://arxiv.org/pdf/1212.0931.pdf) , [Plattner+ (2013)](https://pubs.acs.org/doi/abs/10.1021/ct400355g), …), and investigated briefly in the K=2 case last year by @proteneer .)

Reply:
> Doc request: 1-liner descriptions of each plot.

Added in https://github.com/proteneer/timemachine/pull/1128/commits/be0ef78bddb064b667df991dd04e412451cde6de